content/lectures/lec02.md

```diff
@@ -7,3 +7,190 @@ draft: false
+Status: Downloaded newer image for ubuntu:latest
+docker.io/library/ubuntu:latest
+```
+
+It's important to note here that when we tell Docker to pull the latest verison of Ubuntu, it actually does three seperate pulls from three seperate images. This is becauase Docker images are build inheriently from layers. Each pull is for a seperate layer of the image, and they are built on top of each other to produce the final product.
+
+When you think about it, this is a really smart way to create Docker images. Let's imagine a Docker container that runs Redis, the caching server that acts as a data store. To run Redis, we first need to have a Linux distribution installed, so we can add that as a lower layer of the Docker image which Redis then builds on top of by installing the specific Redis software.
+
+So, now that we have this Ubuntu Docker image installed, what can we do with it? Let's spin up the container and play around with it:
+
+```
+$ docker run -it --name ubuntu-demo ubuntu
+# This will open a terminal within the container running Ubuntu thanks to the "-it" flag
+# Now let's check to make sure we are actually running the latest version of Ubuntu
+root@260a8119abd6:/# cat /etc/os-release
+NAME="Ubuntu"
+VERSION="20.04.1 LTS (Focal Fossa)"
+ID=ubuntu
+ID_LIKE=debian
+PRETTY_NAME="Ubuntu 20.04.1 LTS"
+VERSION_ID="20.04"
+HOME_URL="https://www.ubuntu.com/"
+SUPPORT_URL="https://help.ubuntu.com/"
+BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
+PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
+VERSION_CODENAME=focal
+UBUNTU_CODENAME=focal
+```
+
+We can see that the Docker container is running a fresh install of Ubuntu. Now, we could do all manner of things with this Ubuntu container: install new software, expose ports, connect to a network, run a web server, etc. However, actually doing these things manually within the Ubuntu terminal is _not_ reproducible; we would have to perform those same operations every time we wanted to spin up a new container. Hence, a Dockerfile. When we write a Dockerfile it is essentially a roadmap for how the Docker image should be run, so that all of the installation and configuration of whatever toolchain we need on our container can be written out concisely once and only once.
+
+However, feel free to try out different commands or installing different software on the Ubuntu container. It works the same as any other Linxu command line interface and there is a lot you can do with it. You can type the `exit` comand to leave the Ubuntu terminal at any time.
+
+### Building an App on Docker
+
+Let's build a simple Node web application on Docker to outline the development process. First, create a new directory anywhere on your machine called `node_demo`. Now within this directory will write all of our application code. First we'll start with a Dockerfile that has all of the requisite steps for setting up our Docker application:
```

Comment:
see my comment about putting this in-repo and having them just `cd` into the dir